morebits.js

```diff
@@ -131,14 +131,26 @@ Morebits.pageNameNorm = mw.config.get('wgPageName').replace(/_/g, ' ');
 
 
 /**
- * Create a string for use in regex matching a page name, regardless of the
- * leading character's capitalization.
+ * Create a string for use in regex matching a page name.  Accounts for
+ * leading character's capitalization, underscores as spaces, and special
+ * characters are escaped.
  *
- * @param {string} pageName - Page name without namespace.
- * @returns {string} - For a page name `Foo bar`, returns the string `[Ff]oo bar`.
+ * @param {string} [pageName=`wgTitle`] - Page name without namespace.
+ * Defaults to the current page.
+ * @returns {string} - For a page name `Foo bar`, returns the string `[Ff]oo[_ ]bar`.
+ * @throws If not provided a valid string.
  */
 Morebits.pageNameRegex = function(pageName) {
-	return '[' + pageName[0].toUpperCase() + pageName[0].toLowerCase() + ']' + pageName.slice(1);
+	pageName = pageName || mw.config.get('wgTitle');
+	if (typeof pageName !== 'string') {
+		throw new Error('Not a valid string');
+	}
+	var firstChar = Morebits.string.escapeRegExp(pageName[0]),
+		remainder = Morebits.string.escapeRegExp(pageName.slice(1));
+	if (firstChar.toUpperCase() !== firstChar.toLowerCase()) {
+		return '[' + pageName[0].toUpperCase() + pageName[0].toLowerCase() + ']' + remainder;
```

Comment:
This is still incomplete. `pageName[0].toUpperCase()` should probably be written as `mw.Title.phpCharToUpper(pageName[0])` to account for caps differences between PHP and JS.

Reply:
Cool, didn't know about `phpCharToUpper`!  I almost left a note about this was guaranteed to be imperfect given utf-8, but that probably(?) reduces errors to much much rarer edge cases, that I'm not sure MediaWiki itself handles.

Reply:
Worth doing for `Morebits.string.toUpperCaseFirstChar`, presumably

Reply:
Done, now.  Incorporated the json file for our mocking.

Also, the cases where `escapeRegExp` does something are never cases where the upper and lower differ, so that was unnecessary.

Reply:
Cool. There shouldn't be any edge cases that MW doesn't handle. Any char that gives different uppercases on PHP and JS is included in that file.

Reply:
Yeah, that's what I was saying — anything this would be wrong on MW itself is going to be.